src/Microsoft.Identity.Web.TokenAcquisition/TokenAcquisition.cs

```diff
@@ -585,10 +608,81 @@ private bool IsInvalidClientCertificateOrSignedAssertionError(MsalServiceExcepti
                     _applicationsByAuthorityClientId.TryAdd(GetApplicationKey(mergedOptions), application);
                 }
             }
+            return application;
+        }
+
+        /// <summary>
+        /// Gets a cached ManagedIdentityApplication object or builds a new one if not found.
+        /// </summary>
+        /// <param name="mergedOptions">The configuration options for the app</param>
+        /// <param name="managedIdentityOptions">The configuration specific to managed identity</param>
+        /// <returns>The application object used to request a token with managed identity</returns>
+        internal async Task<IManagedIdentityApplication> GetOrBuildManagedIdentityApplication(
+            MergedOptions mergedOptions, ManagedIdentityOptions managedIdentityOptions)
+        {
+            string key = GetCacheKeyForManagedId(managedIdentityOptions);
+
+            // Check if the application is already built, if so return it without grabbing the key
+            if (_managedIdentityApplicationsByClientId.TryGetValue(key, out IManagedIdentityApplication? application) && application != null)
```

Comment:
I think the way "TryGetValue" type of APIs work is that they return false if the value cannot be fetched. So no need to check if `applicaiton != null`

Reply:
Agreed in this case since I control all values that go into the dictionary and don't allow null ones. However, for future reference, TryGetValue will return false only if the key cannot be found. Even if the value is null for a given key it will return true as long as the key is there.

Reply:
fun fact that `TryGetValue` is faster  when looking up a specific key in a dictionary compared to `ContainsKey`